Abandon messages when an exception occurs in batch function regardless of autocomplete setting

JoshLove-msft opened the pull request:
Previously, we were only abandoning messages in batch functions if an error occurred _and_ AutoCompleteMessages was true. This was inconsistent with the behavior for single-dispatch functions, and with the behavior of the SDK when using the processor directly. The correct behavior is for abandoning to be independent of AutoCompleteMessages.

While working on this fix, an additional two issues were found and fixed:
- Messages received from ReceiveActions were not being abandoned when there was an exception
- Message lock renewal was not being stopped when user handler throws

JoshLove-msft committed:
Abandon messages when an exception occurs in batch function regardless of autocomplete setting

JoshLove-msft commented:
/azp run net - servicebus - tests

azure-sdk commented:
**API change check for `Microsoft.Azure.WebJobs.Extensions.ServiceBus`**

API changes are not detected in this pull request for `Microsoft.Azure.WebJobs.Extensions.ServiceBus`

JoshLove-msft committed:
Fix test

JoshLove-msft commented:
/azp run net - servicebus - tests

jsquire commented on sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/README.md:
```diff
@@ -274,7 +274,9 @@ public static async Task Run(
 ## Troubleshooting
 
-Please refer to [Monitor Azure Functions](https://docs.microsoft.com/azure/azure-functions/functions-monitoring) for troubleshooting guidance.
+If your function triggers an unhandled exception, and you haven't already settled the message, the extension will attempt to abandon the message so that it becomes available for receiving again immediately.
```
```suggestion
If your function triggers an unhandled exception and you haven't already settled the message, the extension will attempt to abandon the message so that it becomes available for receiving again immediately.
```

jsquire commented on sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs:
```diff
@@ -403,42 +403,43 @@ private async Task RunBatchReceiveLoopAsync(CancellationToken cancellationToken)
+                                abandonTasks.Add(receiver.AbandonMessageAsync(message, cancellationToken: CancellationToken.None));
                             }
+
+                            await Task.WhenAll(abandonTasks).ConfigureAwait(false);
```
In the off chance that one of the abandon calls throws and sessions are enabled, would this cause an issue because the block to release the session on L445 will get skipped?   Should we consider wrapping this in a `try/catch` ?

JoshLove-msft replied:
It shouldn't matter because we end up closing the receiver anyway in the fall through catch block.

JoshLove-msft committed:
Fix two issues in SDK:
- Abandon messages received from receiveActions when there is an exception
- Fix bug where message lock renewal is not stopped when user handler throws

JoshLove-msft committed:
remove comma

JoshLove-msft commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ReceiverManager.cs:
```diff
@@ -239,24 +235,25 @@ protected async Task ProcessOneMessageWithinScopeAsync(ServiceBusReceivedMessage
-            ServiceBusReceivedMessage message,
-            CancellationToken processorCancellationToken)
-        {
-            var args = new ProcessMessageEventArgs(
```
Previously the args would not be returned to the caller in case of an exception, which broke the finally block that cancels lock renewal.

JoshLove-msft commented:
/azp run net - servicebus - tests

azure-sdk commented:
**API change check for `Azure.Messaging.ServiceBus`**

API changes are not detected in this pull request for `Azure.Messaging.ServiceBus`

JoshLove-msft committed:
Change log

JoshLove-msft committed:
Fix
